README.md
# Minerva

## A 32-bit RISC-V soft processor

Minerva is a CPU core that currently implements the [RISC-V][1] RV32IM instruction set. Its microarchitecture is described in plain Python code using the [nMigen][2] toolbox.

### Quick start

Minerva currently requires Python 3.6+ and [nMigen][2] on its `master` branch.

    # generate verilog
    python setup.py develop
    python cli.py generate -t v > minerva.v

    # generate verilog with cache
    python cli.py --with-icache generate -t v > minerva.v

    # run some tests
    python setup.py test
    python -m unittest test/test_cache.py

To use Minerva in its minimal configuration, you need to wire the following ports to `minerva_cpu`:

* `clk`
* `rst`
* `ibus__*`
* `dbus__*`
* `external_interrupt`
* `timer_interrupt`
* `software_interrupt`

For an example of Minerva running some code, see 
[here](https://github.com/jfng/minerva-examples).

### Features

The microarchitecture of Minerva is largely inspired by the [LatticeMico32][3] processor.

Minerva is pipelined on 6 stages:

1. **Address**
   The address of the next instruction is calculated and sent to the instruction cache.
2. **Fetch**
   The instruction is read from memory.
3. **Decode**
   The instruction is decoded, and operands are either fetched from the register file or bypassed from the pipeline. Branches are predicted by the static branch predictor.
4. **Execute**
   Simple instructions such as arithmetic and logical operations are completed at this stage.
5. **Memory**
   More complicated instructions such as loads, stores and shifts require a second execution stage.
6. **Writeback**
   Results produced by the instructions are written back to the register file.

![Pipeline Diagram Image](https://docs.google.com/drawings/d/e/2PACX-1vTMkQc8ZJoiJ2AOeFGMkK0QTNx1hSG5wDrG5seLdJ3i61E4ag7wH7VFey44qhvuXotvOKxOw-mFS-VE/pub?w=850&h=761)

The L1 data cache is coupled to a write buffer. Store transactions are in this case done to the write buffer instead of the data bus. This enables stores to proceed in one clock cycle if the buffer isn't full, without having to wait for the bus transaction to complete. Store transactions are then completed in the background as the write buffer gets emptied to the data bus.

### Configuration

The following parameters can be used to configure the Minerva core.

| Parameter         | Default value  | Description                                        |
| ----------------- | -------------- | -------------------------------------------------- |
| `reset_address`   | `0x00000000`   | Reset vector address                               |
| `with_icache`     | `False`        | Enable the instruction cache                       |
| `icache_nways`    | `1`            | Number of ways in the instruction cache            |
| `icache_nlines`   | `128`          | Number of lines in the instruction cache           |
| `icache_nwords`   | `4`            | Number of words in a line of the instruction cache |
| `icache_base`     | `0x00000000`   | Base of the instruction cache address space        |
| `icache_limit`    | `0x80000000`   | Limit of the instruction cache address space       |
| `with_dcache`     | `False`        | Enable the data cache                              |
| `dcache_nways`    | `1`            | Number of ways in the data cache                   |
| `dcache_nlines`   | `128`          | Number of lines in the data cache                  |
| `dcache_nwords`   | `4`            | Number of words in a line of the data cache        |
| `dcache_base`     | `0x00000000`   | Base of the data cache address space               |
| `dcache_limit`    | `0x80000000`   | Limit of the data cache address space              |
| `with_muldiv`     | `False`        | Enable RV32M support                               |
| `with_debug`      | `False`        | Enable the Debug Module                            |
| `with_trigger`    | `False`        | Enable the Trigger Module                          |
| `nb_triggers`     | `8`            | Number of triggers                                 |
| `with_rvfi`       | `False`        | Enable the riscv-formal interface                  |

### Possible improvements

In no particular order:

* RV64I
* Floating Point Unit
* Stateful branch prediction
* MMU
* ...

If you are interested in sponsoring new features or improvements, get in touch at contact [at] lambdaconcept.com .

### License

Minerva is released under the permissive two-clause BSD license.
See LICENSE file for full copyright and license information.

[1]: https://riscv.org/specifications/
[2]: https://github.com/m-labs/nmigen/
[3]: https://github.com/m-labs/lm32/


news.md
 - update install requires - don't think whitequark is still pushing to pypi

 - follow pep setup.py suggestions allowing for easy testing and 
sane imports
   - remove relative imports
   - ```python3 setup.py test``` now works
 - replace dependency of internal nmigen FHDLTestCase using ```nmigen.test.utils``` (which whitequark has already deprecated) - formal tests still passing
 - remove unused signals from L1Cache - formal tests still passing
 - add README link to Minerva "hello world"
 - add "hello world" as test case

cli.py
import argparse
import warnings
from nmigen import cli

from minerva.core import Minerva


def main():
    parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)

    parser.add_argument("--reset-addr",
            type=lambda s: int(s, 16), default="0x00000000",
            help="reset vector address")

    parser.add_argument("--with-icache",
            default=False, action="store_true",
            help="enable the instruction cache")
    parser.add_argument("--with-dcache",
            default=False, action="store_true",
            help="enable the data cache")
    parser.add_argument("--with-muldiv",
            default=False, action="store_true",
            help="enable RV32M support")
    parser.add_argument("--with-debug",
            default=False, action="store_true",
            help="enable the Debug Module")
    parser.add_argument("--with-trigger",
            default=False, action="store_true",
            help="enable the Trigger Module")
    parser.add_argument("--with-rvfi",
            default=False, action="store_true",
            help="enable the riscv-formal interface")

    icache_group = parser.add_argument_group("icache options")
    icache_group.add_argument("--icache-nways",
            type=int, choices=[1, 2], default=1,
            help="number of ways")
    icache_group.add_argument("--icache-nlines",
            type=int, default=32,
            help="number of lines")
    icache_group.add_argument("--icache-nwords",
            type=int, choices=[4, 8, 16], default=4,
            help="number of words in a line")
    icache_group.add_argument("--icache-base",
            type=lambda s: int(s, 16), default="0x00000000",
            help="base address")
    icache_group.add_argument("--icache-limit",
            type=lambda s: int(s, 16), default="0x80000000",
            help="limit address")

    dcache_group = parser.add_argument_group("dcache options")
    dcache_group.add_argument("--dcache-nways",
            type=int, choices=[1, 2], default=1,
            help="number of ways")
    dcache_group.add_argument("--dcache-nlines",
            type=int, default=32,
            help="number of lines")
    dcache_group.add_argument("--dcache-nwords",
            type=int, choices=[4, 8, 16], default=4,
            help="number of words in a line")
    dcache_group.add_argument("--dcache-base",
            type=lambda s: int(s, 16), default="0x00000000",
            help="base address")
    dcache_group.add_argument("--dcache-limit",
            type=lambda s: int(s, 16), default="0x80000000",
            help="limit address")

    trigger_group = parser.add_argument_group("trigger options")
    trigger_group.add_argument("--nb-triggers",
            type=int, default=8,
            help="number of triggers")

    cli.main_parser(parser)

    args = parser.parse_args()

    if args.with_debug and not args.with_trigger:
        warnings.warn("Support for hardware breakpoints requires --with-trigger")

    cpu = Minerva(args.reset_addr,
            args.with_icache, args.icache_nways, args.icache_nlines, args.icache_nwords,
            args.icache_base, args.icache_limit,
            args.with_dcache, args.dcache_nways, args.dcache_nlines, args.dcache_nwords,
            args.dcache_base, args.dcache_limit,
            args.with_muldiv,
            args.with_debug,
            args.with_trigger, args.nb_triggers,
            args.with_rvfi)

    ports = [
        cpu.external_interrupt, cpu.timer_interrupt, cpu.software_interrupt,
        cpu.ibus.ack, cpu.ibus.adr, cpu.ibus.bte, cpu.ibus.cti, cpu.ibus.cyc, cpu.ibus.dat_r,
        cpu.ibus.dat_w, cpu.ibus.sel, cpu.ibus.stb, cpu.ibus.we, cpu.ibus.err,
        cpu.dbus.ack, cpu.dbus.adr, cpu.dbus.bte, cpu.dbus.cti, cpu.dbus.cyc, cpu.dbus.dat_r,
        cpu.dbus.dat_w, cpu.dbus.sel, cpu.dbus.stb, cpu.dbus.we, cpu.dbus.err
    ]

    if args.with_debug:
        ports += [cpu.jtag.tck, cpu.jtag.tdi, cpu.jtag.tdo, cpu.jtag.tms]

    if args.with_rvfi:
        ports += [
            cpu.rvfi.valid, cpu.rvfi.order, cpu.rvfi.insn, cpu.rvfi.trap, cpu.rvfi.halt,
            cpu.rvfi.intr, cpu.rvfi.mode, cpu.rvfi.ixl, cpu.rvfi.rs1_addr, cpu.rvfi.rs2_addr,
            cpu.rvfi.rs1_rdata, cpu.rvfi.rs2_rdata, cpu.rvfi.rd_addr, cpu.rvfi.rd_wdata,
            cpu.rvfi.pc_rdata, cpu.rvfi.pc_wdata, cpu.rvfi.mem_addr, cpu.rvfi.mem_rmask,
            cpu.rvfi.mem_wmask, cpu.rvfi.mem_rdata, cpu.rvfi.mem_wdata
        ]

    cli.main_runner(parser, args, cpu, name="minerva_cpu", ports=ports)


if __name__ == "__main__":
    main()
